test/e2e/instrumentation/monitoring/prometheus.go

```diff
@@ -0,0 +1,385 @@
+	var podAnnotations, serviceAnnotations map[string]string
+	if annotatePods {
+		podAnnotations = map[string]string{
+			"prometheus.io/scrape": "true",
+			"prometheus.io/path":   "/Metrics",
+			"prometheus.io/port":   "8080",
+		}
+	}
+	if annotateService {
+		serviceAnnotations = map[string]string{
+			"prometheus.io/scrape": "true",
+			"prometheus.io/path":   "/Metrics",
+			"prometheus.io/port":   "8080",
+		}
+	}
+	return common.NewMetricExporter(consumerName, f.Namespace.Name, podAnnotations, serviceAnnotations, metricValue, f.ClientSet, f.InternalClientset, f.ScalesGetter)
+}
+
+func prometheusCPUQuery(namespace, podNamePrefix string, rate time.Duration) string {
+	return fmt.Sprintf(`sum(irate(container_cpu_usage_seconds_total{namespace="%v",pod_name=~"%v.*"}[%vm]))`,
+		namespace, podNamePrefix, int64(rate.Minutes()))
+}
+
+func prometheusServiceCustomMetricQuery(namespace, service string) string {
+	return fmt.Sprintf(`sum(QPS{kubernetes_namespace="%v",kubernetes_name="%v"})`, namespace, service)
+}
+
+func prometheusPodCustomMetricQuery(namespace, podNamePrefix string) string {
+	return fmt.Sprintf(`sum(QPS{kubernetes_namespace="%s",kubernetes_pod_name=~"%s.*"})`, namespace, podNamePrefix)
+}
+
+func retryUntil(predicate func() bool, timeout time.Duration) {
+	startTime := time.Now()
+	for {
+		if predicate() {
+			return
+		}
+		if time.Since(startTime) >= testTimeout {
+			break
+		}
+		time.Sleep(prometheusSleepBetweenAttempts)
+	}
+	framework.Failf("monitoring using prometheus test failed")
+}
+
+func allScrapeTargetsAreHealthy(c clientset.Interface) bool {
+	discovery, err := fetchPrometheusTargetDiscovery(c)
+	if err != nil {
+		framework.Logf("failed to query prometheus. Err: %v", err)
+		return false
+	}
+	if len(discovery.ActiveTargets) == 0 {
+		framework.Logf("prometheus is not scraping any targets")
+		return false
+	}
+	for _, target := range discovery.ActiveTargets {
+		if target.Health != HealthGood {
+			framework.Logf("target health not good. Target: %v", target)
+			return false
+		}
+	}
+	return true
+}
+
+func metricAvailableOnNodes(c clientset.Interface, expectedNodesNames []string, metric string, samplesCount int) bool {
+	nodesScrapedTimes, err := getNumberOfTimesMetricScrapedPerNode(c, prometheusRequiredNodesUpDuration, metric)
+	if err != nil {
+		framework.Logf("failed to query prometheus. Err: %v", err)
+		return false
+	}
+	nodesUp := make(map[string]bool)
+	for node, times := range nodesScrapedTimes {
+		if times >= samplesCount {
+			nodesUp[node] = true
+		}
+	}
+
+	if !expectedItemsExist(expectedNodesNames, nodesUp) {
+		framework.Logf("Failed to find all expected Nodes.\n Expected: %v\nActual: %v", expectedNodesNames, nodesUp)
+		return false
+	}
+	return true
+}
+
+func validateQueryValues(c clientset.Interface, query string, expectedValue float64, samplesCount int, errorTolerance float64) bool {
```

Comment:
nit: please stick to some naming convention, e.g. use "validate" prefix in all functions that check if the result is expected or use some other name here ("metricHasCorrectValue"). Personally I'd prefer the latter and use validate.* for functions that do some validation and fail the test/return error if it doesn't pass.